source/common/json/json_sanitizer.cc

```diff
@@ -0,0 +1,77 @@
+#include "source/common/json/json_sanitizer.h"
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_format.h"
+
+namespace Envoy {
+namespace Json {
+
+JsonSanitizer::JsonSanitizer() {
+  // Single-char escape sequences for common control characters.
+  auto control_char = [this](char escape, absl::string_view letter) {
+    char_escapes_[static_cast<uint32_t>(escape)] =
+        std::make_unique<std::string>(absl::StrCat("\\", letter));
+  };
+  control_char('\b', "b");
+  control_char('\f', "f");
+  control_char('\n', "n");
+  control_char('\r', "r");
+  control_char('\t', "t");
+  control_char('\\', "\\");
+  control_char('\"', "\"");
+
+  // Low characters (0-31) not listed above are encoded as unicode 4-digit hex, plus
+  // a few other specific ones.
+  auto unicode_escape = [this](uint32_t index) {
+    if (char_escapes_[index] == nullptr) {
+      char_escapes_[index] = std::make_unique<std::string>(absl::StrFormat("\\u%04x", index));
+    }
+  };
+  unicode_escape('<');
+  unicode_escape('>');
+  unicode_escape(127);
+
+  // Control-characters below 32.
+  for (uint32_t i = 0; i < ' '; ++i) {
+    unicode_escape(i);
+  }
+}
+
+absl::string_view JsonSanitizer::sanitize(std::string& buffer, absl::string_view str) const {
+  size_t past_escape = absl::string_view::npos;
+  for (uint32_t i = 0, n = str.size(); i < n; ++i) {
+    uint32_t index = static_cast<uint32_t>(static_cast<uint8_t>(str[i]));
+    const std::string* escape = char_escapes_[index].get();
+    if (escape != nullptr) {
+      if (past_escape == absl::string_view::npos) {
+        // We only initialize buffer when we first learn we need to add an
+        // escape-sequence to the sanitized string.
+        if (i == 0) {
+          buffer = *escape;
+        } else {
+          buffer = absl::StrCat(str.substr(0, i), *escape);
+        }
+      } else if (i == past_escape) {
+        absl::StrAppend(&buffer, *escape);
+      } else {
+        absl::StrAppend(&buffer, str.substr(past_escape, i - past_escape), *escape);
+      }
+      past_escape = i + 1;
+    }
+  }
```

Comment:
May be codes can be more simple by this way?

```suggestion
  size_t past_escape = 0;
  for (uint32_t i = 0, n = str.size(); i < n; ++i) {
    const std::string* escape = char_escapes_[static_cast<uint8_t>(str[i])].get();
    if (escape != nullptr) {
      absl::StrAppend(&buffer, str.substr(past_escape, i - past_escape), *escape);
      past_escape = i + 1;
    }
  }
```

Reply:
I think the above suggestion is part of the code I propose, but it doesn't handle some of the other cases, such as 2 escaped chars in a row, or clearing the buffer between each call to sanitize that requires escaping. I'll add more comments to make each conditional clearer.

Reply:
The code could be simpler but I wanted to avoid having the user clear `buffer` on every call, because that does not optimize for the case where the string does not require escaping, which in the expected usage will be the common case.

Reply:
Sound reasonable. But string clear just sets string length to zero and first char to '\0', will this brings significant performance impact? 🤔

Reply:
Changing the code to what you suggested definitely made things materially slower, but adding the clear() at the start was only marginally slower (within margin of noise), I guess due to compilers/CPUs doing branch prediction well (clear() will need to do different things depending on whether the std::string has inlined the bytes or note).

I'll mess around a little more with that.

Reply:
Get it. If we can simplify the code without paying a performance price, it is surely best to do so. Otherwise, just ignore my suggestion. 😸